### PR TITLE
Touch Data on Events

### PR DESCRIPTION
**Pull request by kristoferbaxter**
We're lacking tests in this area of the codebase, but we now have support for touch event data in the Worker thread.

### CODE REVIEW

File: src/main-thread/commands/event-subscription.ts [hunk 1]
```diff
@@ -105,6 +126,8 @@ export function EventSubscriptionProcessor(strings: Strings, nodeContext: NodeCo
         [TransferrableKeys.keyCode]: 'keyCode' in event ? event.keyCode : undefined,
         [TransferrableKeys.pageX]: 'pageX' in event ? event.pageX : undefined,
         [TransferrableKeys.pageY]: 'pageY' in event ? event.pageY : undefined,
+        [TransferrableKeys.touches]: createTransferrableTouch(event, 'touches'),
+        [TransferrableKeys.changedTouches]: createTransferrableTouch(event, 'changedTouches'),
```
Comment (choumx) [inline src/main-thread/commands/event-subscription.ts]: Nit: Would be a bit more consistent/readable if the `undefined` condition was extracted here:

```ts
// Counterpart is touchListFromTransfer().
function touchListToTransfer(touchList: TouchList): TransferrableTouchList {
  return Object.values(touchList).map(touch => {
    return [touch.identifier, touch.screenX, ...];
  });
}

[TransferrableKeys.touches]: 'touches' in event ? touchListToTransfer(event.touches) : undefined,
[TransferrableKeys.touches]: 'changedTouches' in event ? touchListToTransfer(event.changedTouches) : undefined,
```

File: src/transfer/TransferrableEvent.ts [hunk 2]
```diff
@@ -17,6 +17,21 @@
...
+interface TransferrableTouch {
+  [key: number]: number;
+  // identifier
+  // screenX
+  // screenY
+  // clientX
+  // clientY
+  // pageX
+  // pageY
+  // target._index_
```
Comment (choumx) [inline src/transfer/TransferrableEvent.ts]: Why are these fields commented out? Shouldn't type checker complain about this?

File: src/worker-thread/Event.ts [hunk 3]
```diff
@@ -80,6 +97,40 @@ const targetFromTransfer = (document: Document, event: TransferrableEvent): Node
...
+        pageX: transferredTouch[5],
+        pageY: transferredTouch[6],
+        target: get(transferredTouch[7] !== 0 ? transferredTouch[7] : document[TransferrableKeys.index]),
+      };
+    });
+
+    return list;
+  }
+  return undefined;
+};
```
Comment (choumx) [inline src/worker-thread/Event.ts]: Ditto for this function and callers below.

File: src/transfer/TransferrableEvent.ts [hunk 4]
```diff
@@ -17,6 +17,21 @@
...
+}
+export interface TransferrableTouchList {
+  [key: number]: TransferrableTouch;
```
Comment (choumx) [inline src/transfer/TransferrableEvent.ts]: Shouldn't interface contain an array of `TransferrableTouch`?

File: src/transfer/TransferrableEvent.ts (hunk 2 above)
Comment (kristoferbaxter) [reply src/transfer/TransferrableEvent.ts]: I'll add a comment, but this is similar to the other Transferables where the index indicates a specific item, and a const enum lets you access specific members in the correct locations without referring to this comment each time.

File: src/transfer/TransferrableEvent.ts (hunk 4 above)
Comment (kristoferbaxter) [reply src/transfer/TransferrableEvent.ts]: We can change the interface here, but these are functionally equivalent for typing.

File: src/main-thread/commands/event-subscription.ts [hunk 5]
```diff
@@ -65,6 +65,27 @@ const fireResizeChange = (workerContext: WorkerContext, cachedWindowSize: [numbe
...
+        touch.identifier,
+        touch.screenX,
+        touch.screenY,
+        touch.clientX,
+        touch.clientY,
+        touch.pageX,
+        touch.pageY,
+        (touch.target as RenderableElement)._index_,
+      ])
+    : undefined;
```
[the remark this answers is not in the document]
Comment (kristoferbaxter) [reply src/main-thread/commands/event-subscription.ts]: Attempting to be consistent with the other methods defined in this module. Perhaps we should tackle them all together in a separate PR?

File: src/main-thread/commands/event-subscription.ts (hunk 1 above)
Comment (kristoferbaxter) [reply src/main-thread/commands/event-subscription.ts]: That's far more consistent with the pattern here. Good call, will change!

File: src/main-thread/commands/event-subscription.ts [hunk 6]
```diff
@@ -65,6 +65,27 @@ const fireResizeChange = (workerContext: WorkerContext, cachedWindowSize: [numbe
     [TransferrableKeys.sync]: cachedWindowSize,
   });
 
+/**
+ * Convert a TouchList into a TransferrableTouchList
+ * @param touches
+ */
+const createTransferrableTouch = (
+  event: Event | KeyboardEvent | MouseEvent | TouchEvent,
```
Comment (dvoytenko) [inline src/main-thread/commands/event-subscription.ts]: A bit lost: why is `KeyboardEvent` here?

File: src/worker-thread/Event.ts [hunk 7]
```diff
@@ -30,6 +30,21 @@ interface EventOptions {
...
+  readonly screenY: number;
+  readonly clientX: number;
+  readonly clientY: number;
+  readonly pageX: number;
+  readonly pageY: number;
+  readonly target: Node | null;
+}
+interface TouchList {
+  [key: number]: Touch;
+  length: number;
```
Comment (dvoytenko) [inline src/worker-thread/Event.ts]: Isn't this one of those APIs that also have an `item()` method? I think they are still used quite a bit, despite the `[i]` accessor.

File: src/main-thread/commands/event-subscription.ts (hunk 6 above)
Comment (kristoferbaxter) [reply src/main-thread/commands/event-subscription.ts]: Fixed. Thanks!

File: src/worker-thread/Event.ts (hunk 7 above)
Comment (kristoferbaxter) [reply src/worker-thread/Event.ts]: Added `item(index: number) => Touch | null` method.

**Pull request merged**